File: Semestr 3/Paradygmaty_Programowania/Ocaml/cw_5.ml
type 'a llist = 
	LNil 
	| LCons of 'a * (unit -> 'a llist);;

let ltl = function
	| LNil -> failwith "ltl"
	| LCons (_, xf) -> xf();;

let rec ltake = function
	| (0, _) -> []
	| (_, LNil) -> []
	| (n, LCons(x,xf)) -> x::ltake(n-1, xf());;

let rec lfrom k = LCons (k, function () -> lfrom (k+1));;

let rec toLazyList = function
	| [] -> LNil
	| h::t -> LCons(h, function () -> toLazyList t);;

let rec (@$) ll1 ll2 =
	match ll1 with
		| LNil -> ll2
		| LCons(x, xf) -> LCons(x, function () -> (xf()) @$ ll2);;

(* zadanie 1 *)
let lrepeat k lli =
	let rec rep m llista =
		match m, llista with
		| _, LNil -> LNil
		| 0, LCons(x, xf) -> rep k (xf())
		| n, LCons(x, _) -> LCons(x, ( function () -> rep (n-1) llista))
	in rep k lli;;

(* zadanie 2 *)
let lfib = 
	let rec fibo (a,b) = LCons(a, function () -> fibo(b,a+b))
	in fibo(0,1);;
	

(* zadanie 3 *)
type 'a lBT = 
	LEmpty 
	| LNode of 'a * (unit ->'a lBT) * (unit -> 'a lBT);;


let rec lTree n = 
	LNode (n, (function() -> lTree(2*n)), (function() -> lTree(2*n+1)));;

let breadth lbt =
	let rec bb = function 
		| [] -> LNil
		| LNode (v, l, r)::t -> LCons(v,fun () -> bb(t@[l();r()]))
		| LEmpty::t -> bb t
	in bb [lbt];;

  
	

File: Semestr 3/Paradygmaty_Programowania/Scala/src/cw_5.scala
// Zadanie 1

def lrepeat[A] (k:Int,s:Stream[A]) = {
  def rep[A] (m:Int, llista:Stream[A]):Stream[A] =
    (m, llista) match {
    case (_, Stream.Empty) => Stream.empty
    case (0, h#::xs) => rep (k, xs)
    case (n, h#::xs) => Stream.cons(h, rep (n-1, llista))
  }
  rep (k, s)
}

// zadanie 2
def fibo(a:BigInt,b:BigInt):Stream[BigInt] =
  Stream.cons(a, fibo(b, a+b))

val lfib = {
  def fibo(a:BigInt,b:BigInt):Stream[BigInt] ={
    Stream.cons(a, fibo(b, a+b))}
  fibo(0,1)
}

// zadanie 3

sealed trait lBT[+A]
case object LEmpty extends lBT[Nothing]
case class LNode[+A](elem:A, left:()=>lBT[A], right:()=>lBT[A]) extends lBT[A]

def lTree (n:Int):lBT[Int] =
  LNode(n, () => lTree(2*n), ()=>lTree(2*n+1))

def breadth[A] (lbt:lBT[A]) = {
    def bb (queue:List[lBT[A]]):Stream[A] = 
        queue match {            
            case Nil => Stream.empty
            case LNode(v, lt, rt)::t => Stream.cons(v,bb(t++List(lt(), rt())))
            case LEmpty::t => bb(t)
       }
    bb (List(lbt))
}  
